common/k8s/src/event_source.rs

```diff
@@ -0,0 +1,408 @@
+                    .map(|age| {
+                        age.num_seconds()
+                            .try_into()
+                            .map_err(|_| warn!("age too large, could not fit {} into i32", age))
+                            .ok()
+                    })
+                    .flatten(),
+                count,
+            },
+        }
+    }
+}
+
+// TODO test from...
+
+#[derive(Serialize, Debug)]
+struct EventLogLine {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    _name: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    level: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    message: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    _app: Option<String>,
+    kube: EventLogLineInner,
+}
+
+#[derive(Serialize, Debug)]
+struct EventLogLineInner {
+    type_: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    action: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    resource: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    name: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    namespace: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    reason: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    message: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    component: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    node: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    first_time: Option<Time>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    time: Option<Time>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    age: Option<i32>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    count: Option<i32>,
+}
+
+pub struct K8sEventStream {
+    client: Client,
+    pod_name: Option<String>,
+    namespace: Option<String>,
+}
+
+pub enum StreamElem<T> {
+    Waiting,
+    Event(T),
+}
+
+impl K8sEventStream {
+    pub fn new(
+        config: kube::Config,
+        pod_name: Option<String>,
+        namespace: Option<String>,
+    ) -> Result<Self, K8sError> {
+        Ok(Self {
+            client: Client::new(config),
+            pod_name,
+            namespace,
+        })
+    }
+
+    pub fn try_default(
+        pod_name: Option<String>,
+        namespace: Option<String>,
+    ) -> Result<Self, K8sError> {
+        let config = match Config::from_cluster_env() {
+            Ok(v) => v,
+            Err(e) => {
+                return Err(K8sError::InitializationError(format!(
+                    "unable to get cluster configuration info: {}",
+                    e
+                )))
+            }
+        };
+        Self::new(config, pod_name, namespace)
+    }
+
+    pub async fn event_stream(self) -> Result<impl Stream<Item = Vec<LineBuilder>> + Send, String> {
```

Comment:
This is in dire need or being split up, but this is a first working version.

The general idea is that there are a couple of streams, on that will keep producing "wait" until the agent needs to start logging events, at which point the "wait" stream will end, and the event logging stream can start.